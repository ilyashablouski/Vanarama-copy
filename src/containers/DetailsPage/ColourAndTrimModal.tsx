import React, { useState } from 'react';
import ImageV2 from 'core/atoms/image/ImageV2';
import Price from 'core/atoms/price';
import Button from 'core/atoms/button/Button';
import Text from 'core/atoms/text/Text';
import Icon from 'core/atoms/icon';
import Close from 'core/assets/icons/Close';
import Heading from 'core/atoms/heading';
import Tab from 'core/molecules/tabs/Tab';
import TabList from 'core/molecules/tabs/TabList';
import Tabs from 'core/molecules/tabs';
import ColourTrimColumn from 'core/atoms/colour-trim-column/ColourTrimColumn';
import { IOptionsList } from '../../types/detailsPage';
import { Nullable } from '../../types/common';

interface IColourAndTrimModalProps {
  price: number;
  toggleColorAndTrimModalVisible: () => void;
  headingText: string;
  sortedTrimList: Nullable<IOptionsList[]>;
  colourData: Nullable<IOptionsList[]>;
  isMobile: boolean;
  selectedColour: Nullable<number>;
  setSelectedColour: React.Dispatch<React.SetStateAction<number | null>>;
  selectedTrim: Nullable<number>;
  setSelectedTrim: React.Dispatch<React.SetStateAction<number | null>>;
  setIsFactoryOrder: React.Dispatch<React.SetStateAction<boolean | undefined>>;
  imageUrl: string;
  manufacturerName: string;
}

const MOBILE_TABS = [
  {
    tabName: 'Colour',
    index: 1,
  },
  {
    tabName: 'Trim',
    index: 2,
  },
];

const ColourAndTrimModal: React.FC<IColourAndTrimModalProps> = ({
  price,
  toggleColorAndTrimModalVisible,
  headingText,
  colourData,
  isMobile,
  selectedColour,
  setSelectedColour,
  selectedTrim,
  setSelectedTrim,
  sortedTrimList,
  setIsFactoryOrder,
  imageUrl,
  manufacturerName,
}) => {
  const [activeTab, setActiveTab] = useState(1);

  const changeColour = (
    optionId: number,
    isFactoryOrder: boolean | undefined,
  ) => {
    setSelectedColour(optionId);
    setIsFactoryOrder(isFactoryOrder);
  };
  const changeTrim = (optionId: number) => {
    setSelectedTrim(optionId);
  };

  return (
    <>
      <div className="color-trim">
        <div className="color-trim-header">
          <Text color="black" size="large" className="-b">
            Choose Your Color & Trim
          </Text>
          <Icon
            dataTestId="icon-close"
            icon={<Close />}
            onClick={toggleColorAndTrimModalVisible}
          />
        </div>

        <div className="-container -mb-400">
          <ImageV2
            quality={60}
            objectFit="cover"
            lazyLoad
            src={imageUrl}
            alt={manufacturerName}
            plain
          />
        </div>

        {isMobile && (
          <Tabs
            align="expand"
            activeIndex={activeTab}
            onChange={index => setActiveTab(index)}
            className="-mb-400"
          >
            <TabList>
              {MOBILE_TABS.map(tab => (
                <Tab index={tab.index} key={tab.tabName} className="-p-400">
                  {tab.tabName}
                </Tab>
              ))}
            </TabList>
          </Tabs>
        )}

        {((!isMobile && colourData) ||
          (isMobile && activeTab === 1 && colourData)) && (
          <div className="color-trim-box">
            <Heading
              tag="div"
              size="regular"
              color="black"
              className="headingText -mb-500"
            >
              PAINT COLOUR
            </Heading>
            <div className="row:cards-3col">
              {colourData?.map(item => {
                return (
                  <ColourTrimColumn
                    item={item}
                    selectedItem={selectedColour}
                    setSelectedItem={changeColour}
                    key={item.leadTime}
                  />
                );
              })}
            </div>
          </div>
        )}

        {((!isMobile && sortedTrimList?.length) ||
          (isMobile && activeTab === 2 && sortedTrimList?.length)) && (
          <div className="color-trim-box">
            <Heading
              tag="div"
              size="regular"
              color="black"
              className="headingText -mb-500"
            >
              TRIM
            </Heading>
            <div className="row:cards-3col">
              {sortedTrimList?.map(item => {
                return (
                  <ColourTrimColumn
                    item={item}
                    selectedItem={selectedTrim}
                    setSelectedItem={changeTrim}
                    key={item.leadTime}
                  />
                );
              })}
            </div>
          </div>
        )}
      </div>
      <div className="color-trim-footer">
        <div>
          <Price price={price} size="large" />
          <Heading
            tag="div"
            size="small"
            color="darker"
            className="headingText -pt-100"
          >
            {headingText}
          </Heading>
        </div>
        <Button
          label="Apply"
          color="teal"
          onClick={toggleColorAndTrimModalVisible}
          size="regular"
        />
      </div>
    </>
  );
};

export default ColourAndTrimModal;
