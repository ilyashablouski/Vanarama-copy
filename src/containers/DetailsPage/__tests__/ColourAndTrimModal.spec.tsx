import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import React from 'react';
import ColourAndTrimModal from '../ColourAndTrimModal';

const toggleColorAndTrimModalVisible = jest.fn();
const setSelectedColour = jest.fn();
const setSelectedTrim = jest.fn();
const setIsFactoryOrder = jest.fn();

const mock = {
  price: 123,
  toggleColorAndTrimModalVisible,
  headingText: 'text',
  sortedTrimList: [
    {
      leadTime: '14-21 Day Delivery',
      options: [
        {
          label: 'Leather - Cranberry red',
          optionId: 104562,
          hotOffer: false,
        },
      ],
    },
    {
      leadTime: 'Factory Order',
      options: [
        {
          label: 'Blue',
          optionId: 134567,
          hotOffer: false,
        },
      ],
    },
  ],
  colourData: [
    {
      leadTime: '8-12 Week Delivery',
      hotOffer: true,
      options: [
        {
          hotOffer: true,
          label: 'Metallic - Infra Red',
          optionId: 89191,
          hex: 'b50c0f',
        },
      ],
    },
    {
      leadTime: 'Factory Order',
      hotOffer: true,
      options: [
        {
          hotOffer: true,
          label: 'Premium paint - Blue flame',
          optionId: 140521,
          hex: '407cd6',
        },
      ],
    },
  ],
  isMobile: false,
  selectedColour: 89191,
  setSelectedColour,
  selectedTrim: 104562,
  setSelectedTrim,
  setIsFactoryOrder,
  imacaAssets: null,
  isCar: true,
};

describe('<ColourAndTrimModal>', () => {
  it('should render correctly', async () => {
    const getComponent = render(<ColourAndTrimModal {...mock} />);

    await waitFor(() => {
      expect(screen.getAllByTestId(`HOT OFFER`)[0]).toBeInTheDocument();
    });

    const tree = getComponent.baseElement;
    expect(tree).toMatchSnapshot();
  });

  it('should close', async () => {
    render(<ColourAndTrimModal {...mock} />);

    await waitFor(() => {
      expect(screen.getAllByTestId(`HOT OFFER`)[0]).toBeInTheDocument();
    });

    fireEvent.click(screen.getByTestId('icon-close'));
    expect(toggleColorAndTrimModalVisible).toBeCalled();

    fireEvent.click(screen.getByText('Apply'));
    expect(toggleColorAndTrimModalVisible).toBeCalled();
  });

  it('should change colour', async () => {
    render(<ColourAndTrimModal {...mock} />);

    await waitFor(() => {
      expect(screen.getAllByTestId(`HOT OFFER`)[0]).toBeInTheDocument();
    });

    fireEvent.click(screen.getByTestId('140521'));
    expect(setIsFactoryOrder).toBeCalledWith(true);
    expect(setSelectedColour).toBeCalledWith(140521);
  });

  it('should change trim', async () => {
    render(<ColourAndTrimModal {...mock} />);

    await waitFor(() => {
      expect(screen.getAllByTestId(`HOT OFFER`)[0]).toBeInTheDocument();
    });

    fireEvent.click(screen.getByTestId('134567'));
    expect(setIsFactoryOrder).not.toBeCalled();
    expect(setSelectedTrim).toBeCalledWith(134567);
  });
});
