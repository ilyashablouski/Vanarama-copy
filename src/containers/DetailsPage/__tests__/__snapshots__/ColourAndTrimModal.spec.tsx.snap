// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColourAndTrimModal> should render correctly 1`] = `
<body>
  <div>
    <div
      class="color-trim"
    >
      <div
        class="color-trim-header"
      >
        <span
          class="text -b -large -black"
        >
          Choose Your Color & Trim
        </span>
        <i
          class="icon"
          data-testid="icon-close"
        >
          <svg
            height="1em"
            viewBox="0 0 512 512"
            width="1em"
          >
            <path
              d="M289.94,256l95-95A24,24,0,0,0,351,127l-95,95-95-95A24,24,0,0,0,127,161l95,95-95,95A24,24,0,1,0,161,385l95-95,95,95A24,24,0,0,0,385,351Z"
            />
          </svg>
        </i>
      </div>
      <div
        class="-container -mb-400"
      >
        <div
          class="image -expand -plain"
        >
          <span
            style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
          >
            <img
              alt="manufacturer"
              class="image--native"
              data-nimg="fill"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: cover;"
            />
            <noscript />
          </span>
        </div>
      </div>
      <div
        class="color-trim-box"
      >
        <div
          class="heading headingText -mb-500 -black -regular"
        >
          PAINT COLOUR
        </div>
        <div
          class="row:cards-3col"
        >
          <div>
            <span
              class="heading -mb-400 -black -small"
            >
              8-12 Week Delivery
            </span>
            <div
              class="row:cards-1col"
            >
              <input
                checked=""
                class="choice-input visually-hidden"
                data-testid="89191"
                id="Metallic - Infra Red"
                name="8-12 Week Delivery"
                type="radio"
                value="Metallic - Infra Red"
              />
              <label
                class="colour-trim-checkbox active"
                for="Metallic - Infra Red"
              >
                <span
                  class="text -b -mb-100 -small -orange"
                  data-testid="HOT OFFER"
                >
                  <i
                    class="icon -orange"
                  >
                    <svg
                      height="1em"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M394.23,197.56a300.43,300.43,0,0,0-53.37-90C301.2,61.65,249.05,32,208,32a16,16,0,0,0-15.48,20c13.87,53-14.88,97.07-45.31,143.72C122,234.36,96,274.27,96,320c0,88.22,71.78,160,160,160s160-71.78,160-160C416,276.7,408.68,235.51,394.23,197.56ZM288.33,418.69C278,429.69,265.05,432,256,432s-22-2.31-32.33-13.31S208,390.24,208,368c0-25.14,8.82-44.28,17.34-62.78,4.95-10.74,10-21.67,13-33.37a8,8,0,0,1,12.49-4.51A126.48,126.48,0,0,1,275,292c18.17,24,29,52.42,29,76C304,390.24,298.58,407.77,288.33,418.69Z"
                      />
                    </svg>
                  </i>
                  HOT OFFER
                </span>
                Metallic - Infra Red
                <div
                  class="colour-box"
                  style="background-color: rgb(181, 12, 15);"
                />
              </label>
            </div>
          </div>
          <div>
            <span
              class="heading -mb-400 -black -small"
            >
              Factory Order
            </span>
            <div
              class="row:cards-1col"
            >
              <input
                class="choice-input visually-hidden"
                data-testid="140521"
                id="Premium paint - Blue flame"
                name="Factory Order"
                type="radio"
                value="Premium paint - Blue flame"
              />
              <label
                class="colour-trim-checkbox"
                for="Premium paint - Blue flame"
              >
                <span
                  class="text -b -mb-100 -small -orange"
                  data-testid="HOT OFFER"
                >
                  <i
                    class="icon -orange"
                  >
                    <svg
                      height="1em"
                      viewBox="0 0 512 512"
                      width="1em"
                    >
                      <path
                        d="M394.23,197.56a300.43,300.43,0,0,0-53.37-90C301.2,61.65,249.05,32,208,32a16,16,0,0,0-15.48,20c13.87,53-14.88,97.07-45.31,143.72C122,234.36,96,274.27,96,320c0,88.22,71.78,160,160,160s160-71.78,160-160C416,276.7,408.68,235.51,394.23,197.56ZM288.33,418.69C278,429.69,265.05,432,256,432s-22-2.31-32.33-13.31S208,390.24,208,368c0-25.14,8.82-44.28,17.34-62.78,4.95-10.74,10-21.67,13-33.37a8,8,0,0,1,12.49-4.51A126.48,126.48,0,0,1,275,292c18.17,24,29,52.42,29,76C304,390.24,298.58,407.77,288.33,418.69Z"
                      />
                    </svg>
                  </i>
                  HOT OFFER
                </span>
                Premium paint - Blue flame
                <div
                  class="colour-box"
                  style="background-color: rgb(64, 124, 214);"
                />
              </label>
            </div>
          </div>
        </div>
      </div>
      <div
        class="color-trim-box"
      >
        <div
          class="heading headingText -mb-500 -black -regular"
        >
          TRIM
        </div>
        <div
          class="row:cards-3col"
        >
          <div>
            <span
              class="heading -mb-400 -black -small"
            >
              14-21 Day Delivery
            </span>
            <div
              class="row:cards-1col"
            >
              <input
                checked=""
                class="choice-input visually-hidden"
                data-testid="104562"
                id="Leather - Cranberry red"
                name="14-21 Day Delivery"
                type="radio"
                value="Leather - Cranberry red"
              />
              <label
                class="colour-trim-checkbox active"
                for="Leather - Cranberry red"
              >
                Leather - Cranberry red
              </label>
            </div>
          </div>
          <div>
            <span
              class="heading -mb-400 -black -small"
            >
              Factory Order
            </span>
            <div
              class="row:cards-1col"
            >
              <input
                class="choice-input visually-hidden"
                data-testid="134567"
                id="Blue"
                name="Factory Order"
                type="radio"
                value="Blue"
              />
              <label
                class="colour-trim-checkbox"
                for="Blue"
              >
                Blue
              </label>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="color-trim-footer"
    >
      <div>
        <div
          class="price -large -black"
        >
          <div
            class="price--inner"
          >
            <span
              class="price--sub"
            >
              Â£
            </span>
            <span
              class="price--pounds"
            >
              123
            </span>
            <span
              class="price--sub"
            >
              .
              00
            </span>
          </div>
        </div>
        <div
          class="heading headingText -pt-100 -darker -small"
        >
          text
        </div>
      </div>
      <button
        class="button -teal -regular -solid"
      >
        <div
          class="button--inner"
        >
          Apply
        </div>
      </button>
    </div>
  </div>
</body>
`;
